### ISSUE TITLE
simpsum bug

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#683]
Imported from SourceForge on 2024-07-03 07:35:00
Created by **[*anonymous](https://sourceforge.net/u/*anonymous/)** on 2002-10-18 15:18:23
Original: https://sourceforge.net/p/maxima/bugs/164

---

Hi\!

Unfortunately I have found a bug in simpsum \(it seems\):

\(C1\)
'SUM\(BINOMIAL\(2,2-k\)-BINOMIAL\(2,1-k\),k,1,2\),simpsum;

\(D1\)                                   3

\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\* wrong \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*

\(C2\) 'SUM\(BINOMIAL\(2,2-k\)-BINOMIAL\(2,1-k\),k,1,2\),sum;

\(D2\)                                   2

\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\* correct \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*

\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\* however : \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*

\(C3\)
'SUM\(BINOMIAL\(x,2-k\)-BINOMIAL\(x,1-k\),k,1,2\),simpsum;

\(D3\)                                   x

\(C4\) 'SUM\(BINOMIAL\(x,2-k\)-BINOMIAL\(x,1-k\),k,1,2\),sum;

\(D4\)                                   x
\(C5\) bug\_report\(\);

Maxima version: 5.9.0rc1
Maxima build date: 11:40 9/3/2002
host type: i686-pc-linux-gnu
lisp-implementation-type: Kyoto Common Lisp
lisp-implementation-version: GCL-2-5.0

Martin

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 07:35:01
Created by **[kratt5](https://sourceforge.net/u/kratt5/)** on 2002-11-26 13:49:18
Original: https://sourceforge.net/p/maxima/bugs/164/#cbf5

---

Logged In: YES 
user\_id=651552

OK, I got the bug now:

The fix is rather obvious \(except that I had to dig through
all of sum,
see below\) and the bug is rather serious \(I can produce lots of
everyday examples,

\(C1\) sum\(f\(k\)+1,k,1,n\),simpsum;

\(D2\)                                   n

so I propose to include it in 5.9.0 \!\!\!

I didn't change some things I would like to, I think this is
for after
5.9.0...

fix and explanation below
Martin

\-----------------------------fix-------------------------------------
diff combin.lisp combin.lisp.~1.2.~
915,920d914
&amp;lt; ; Kratt5, 26.11.2002
&amp;lt; ; adsum's and adusum's the sum of e.
&amp;lt;
&amp;lt; ; It's result is discarded. \(at least I think that this
function is
&amp;lt; ; only called by sumsum, but there are lots of places
where a variable
&amp;lt; ; is called &amp;quot;sum&amp;quot;...\)
930,939d923
&amp;lt; ; this is to deal with linearity Kratt5, 26.11.2002
&amp;lt;       \(\(let \(\*a \*n \(var \*var\*\)\)
&amp;lt;          \(cond \(\(prog2 \(m2 e '\(\(mtimes\) \(\(coefftt\) \(var\*
\(set\) \*a freevar\)\)
&amp;lt;                                         \(\(coefftt\) \(var\*
\(set\) \*n true\)\)\)
&amp;lt;                            nil\)
&amp;lt;                     \(not \(equal \*a 1\)\)\)
&amp;lt; ;; we have to return T, so that sum is exited if the test
was successful
&amp;lt;                 \(prog2 \(sum \*n \(list '\(mtimes\) y \*a\)\)
&amp;lt;                     T\)\)\)\)\)
&amp;lt; ;;
943,944c927
&amp;lt;        \#+cl \(adusum \(list '\(mtimes\) e y\)\) ;; Kratt5 26.11.2002
&amp;lt; ;;       nil
\---
&amp;gt;        nil

\--------------------------end
fix------------------------------------

I tested it with

Maxima version: 5.9.0rc3
Maxima build date: 13:52 11/18/2002
host type: i686-pc-linux-gnu
lisp-implementation-type: Kyoto Common Lisp
lisp-implementation-version: GCL-2-5.0

\----------------------------explanation ------------------
\(lisp level\)

the structure of $sum is roughly as follows:

$sum: argcheck, call dosum with meval'd bounds

dosum: didn't look at this too much

after this, meval calls simpsum

\*\*\*\* if you type 'sum\(f\(k\),k,1,n\),simpsum; meval calls only
simpsum \*\*\*\*

$simpsum: call simpsum1

simpsum1: checks lo=hi, otherwise
exp not depending on the summation index, otherwise
if $simpsum, call simpsum2

\*\*\*\* simpsum2 is found in combin.lisp \*\*\*\*

simpsum2: sets up a variable \*plus, which will contain
all the stuff
which is added together at the end
calls sumsum

sumsum: returns the part of the expression it was able
to sum up
\(this is the contents of the variable &amp;quot;usum&amp;quot;\),
all the rest
\(the contents of the variable &amp;quot;sum&amp;quot;\) is put
into the
variable \*plus, which is then used by simpsum2
calls sum

sum: adsum's and adusum's the sum of e.
It's result is discarded. \(at least I think
that this
function is only called by sumsum, but there
are lots of
places where a variable is called &amp;quot;sum&amp;quot;...\)

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 07:35:05
Created by **[kratt5](https://sourceforge.net/u/kratt5/)** on 2002-11-26 19:04:43
Original: https://sourceforge.net/p/maxima/bugs/164/#25d8

---

Logged In: YES 
user\_id=651552

&amp;gt; Can you explain why there is a conditionalization
&amp;gt; for \#+cl there?  I don't see any reason for this
&amp;gt; code to depend on common lisp or not.  My thought
&amp;gt; on fixing the code previously displayed was to
&amp;gt; just remove \#-cl
&amp;gt; RJF

Yes, I can explain it. It's there because I was very stupid.
Here is the
right fix:
\(and THANK YOU, I feel a little ashamed...\)

diff combin.lisp combin.lisp.~1.2.~
915,920d914
&amp;lt; ; Kratt5, 26.11.2002
&amp;lt; ; adsum's and adusum's the sum of e.
&amp;lt;
&amp;lt; ; It's result is discarded. \(at least I think that this
function is
&amp;lt; ; only called by sumsum, but there are lots of places
where a variable
&amp;lt; ; is called &amp;quot;sum&amp;quot;...\)
933,935c927,929
&amp;lt; ;;       nil    ;; Kratt5 26.11.2002
&amp;lt; ;      \#-cl
&amp;lt;        \(let \(\*a \*n\) \(var \*var\*\)\) ; freevar expects &amp;quot;var&amp;quot;,
not &amp;quot;\*var\*&amp;quot;
\---
&amp;gt;        nil
&amp;gt;        \#-cl
&amp;gt;        \(let \(\*a \*n\)

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 07:35:08
Created by **[kratt5](https://sourceforge.net/u/kratt5/)** on 2002-11-26 19:16:10
Original: https://sourceforge.net/p/maxima/bugs/164/#64ff

---

Logged In: YES 
user\_id=651552

sorry, below was a typo \(a parenthesis didn't get deleted\)

933,935c927,929
&amp;lt; ;;       nil    ;; Kratt5 26.11.2002
&amp;lt; ;      \#-cl
&amp;lt;        \(let \(\*a \*n \(var \*var\*\)\) ; freevar expects &amp;quot;var&amp;quot;,
not &amp;quot;\*var\*&amp;quot;
\---
&amp;gt;        nil
&amp;gt;        \#-cl
&amp;gt;        \(let \(\*a \*n\)

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 07:35:15
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2005-11-24 08:39:44
Original: https://sourceforge.net/p/maxima/bugs/164/#84dd

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 07:35:19
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2005-11-25 15:51:24
Original: https://sourceforge.net/p/maxima/bugs/164/#50cb

---

Logged In: YES 
user\_id=501686

Reopening this report because the example shown below
\(sum\(f\(k\)+1,k,1,n\),simpsum;\) is still present. Not clear to
me whether any of the patches shown below was ever applied.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 07:35:23
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2005-11-25 15:51:24
Original: https://sourceforge.net/p/maxima/bugs/164/#bc2a

---

- **status**: closed --> open
- **assigned_to**: nobody --> robert_dodier

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 07:35:26
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2005-12-02 06:34:15
Original: https://sourceforge.net/p/maxima/bugs/164/#0fc3

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 07:35:30
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2005-12-02 06:34:15
Original: https://sourceforge.net/p/maxima/bugs/164/#461c

---

Logged In: YES 
user\_id=501686

sum\(f\(k\)+1,k,1,n\),simpsum;  =&gt;  n+'sum\(f\(k\),k,1,n\)
now, due to r1.10 src/combin.lisp.
\(need to call ADUSUM in SUM to add up parts
of summand which depend on summation index.\)
I wasn't able to decipher the patches stated below, so I
figured out a different patch, which works for the examples
which I added to tests/rtestsum.mac. Closing this bug as fixed.
